Add self-start procedure to CCE

moxcodes opened the pull request:
## Proposed changes

Previously, CCE did not participate in self-start. We seem to have 'gotten away' with it in cases of multistep methods on the rare standalone run that used it, but it's definitely not the right way to do things.
This makes a few tweaks to the self start procedure necessary to support CCE, and includes the self start procedure in the CCE actions.

### Upgrade instructions

<!-- UPGRADE INSTRUCTIONS -->
Any changes to the CCE initialization procedure involving the time step will likely need to be rethought in light of these changes.
<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

moxcodes committed:
Add self-start to CCE

moxcodes committed:
Allow multiple histories in `SelfStart`

Sometimes (particularly in CCE), we need to store multiple histories because the
underlying variable types differ.
This change permits using multiple histories during self-start, under the
constraint that the integration order of the histories must match.

moxcodes committed:
Consistently use System template parameter in SelfStart actions

moxcodes committed:
Allow template template packs in `tt::is_a_lambda`

moxcodes committed:
Allow non-monotonic steps in RT analytic solution for self-start

But also prohibit using substep methods, because that will cause
performance problems

wthrowe commented on src/Time/Actions/SelfStartActions.hpp:
```diff
@@ -113,18 +113,26 @@ namespace Actions {
+
+template <typename TagList>
+struct get_all_history_tags {
+  using type =
```
Is there a reason for this to be a lazy metafunction rather than an eager one?

wthrowe commented on tests/Unit/Time/Actions/Test_SelfStartActions.cpp:
```diff
@@ -205,33 +250,36 @@ using not_self_start_action = std::negation<std::disjunction<
 // Returns true if the last action jumped.
-template <typename Stop, typename Whitelist, bool HasPrimitives>
+template <typename Stop, typename Whitelist, bool MultipleHistories = false,
+          bool HasPrimitives>
```
Default isn't useful here.  If `HasPrimitives` can be deduced, then so can `MultipleHistories`.

wthrowe commented on src/Utilities/TypeTraits/IsA.hpp:
```diff
@@ -55,16 +55,16 @@ using is_a_t = typename is_a<U, Args...>::type;
 /// @}
```
Typo in commit message: "`tt::is_a` lambda" -> "`tt::is_a_lambda`"

wthrowe commented on tests/Unit/Evolution/Systems/Cce/Actions/Test_AnalyticBoundaryCommunication.cpp:
```diff
@@ -170,8 +170,9 @@ SPECTRE_TEST_CASE(
   ActionTesting::MockRuntimeSystem<test_metavariables> runner{
       tuples::tagged_tuple_from_typelist<
           Parallel::get_const_global_cache_tags<test_metavariables>>{
-          l_max, end_time, start_time, number_of_radial_points,
-          std::make_unique<::TimeSteppers::DormandPrince5>()}};
+          l_max, end_time, start_time,
+          std::make_unique<::TimeSteppers::DormandPrince5>(),
+          number_of_radial_points}};
```
For my understanding, this is because of the change in `InitializeWorldtubeBoundary::const_global_cache_tags`, correct?

moxcodes replied:
Yes, that's right -- in this case, the runtime system needed the time stepper regardless, but adding it to `InitializeWorldtubeBoundary::const_global_cache_tags` reordered the tag list.

wthrowe commented on src/Evolution/Systems/Cce/Actions/InitializeFirstHypersurface.hpp:
```diff
@@ -54,6 +54,17 @@ struct InitializeFirstHypersurface {
+        db::get<::Tags::TimeStepId>(box)
+                .substep_time()
+                .fraction()
+                .numerator() != 0) {
```
I think just `.fraction() != 0` is clearer.

wthrowe commented on src/Evolution/Systems/Cce/AnalyticBoundaryDataManager.hpp:
```diff
@@ -68,6 +68,11 @@ class AnalyticBoundaryDataManager {
 
   size_t get_l_max() const noexcept { return l_max_; }
 
+  const std::unique_ptr<Solutions::WorldtubeData>& get_generator()
```
Do you actually need the `unique_ptr`, or can you just return a reference to the contained value?

Also, I think this should be in the "Allow non-monotonic steps in RT analytic solution for self-start" commit.

moxcodes commented:
fixups posted.
Cheers!

wthrowe commented:
Fixups look fine, but apparently something went wrong because some configurations are segfaulting now.

moxcodes commented:
fixups squashed in, along with a fix for the segfault.
The mistake was I used the `&` operator on the returned lvalue ref -- I need to use `std::addressof` for that. The diff for that change (applied to `Allow non-monotonic...` commit) is:
```
--- a/src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp
+++ b/src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp
@@ -60,9 +60,9 @@ struct InitializeWorldtubeBoundaryBase {
                     const ParallelComponent* const /*meta*/) noexcept {
     if constexpr (std::is_same_v<Tags::AnalyticBoundaryDataManager,
                                  tmpl::front<ManagerTags>>) {
-      if (dynamic_cast<const Solutions::RobinsonTrautman*>(
-              &(db::get<Tags::AnalyticBoundaryDataManager>(box)
-                    .get_generator())) != nullptr) {
+      if (dynamic_cast<const Solutions::RobinsonTrautman*>(std::addressof(
+              (db::get<Tags::AnalyticBoundaryDataManager>(box)
+               .get_generator()))) != nullptr) {
         if(db::get<::Tags::TimeStepper<>>(box).number_of_substeps() != 1) {
           ERROR(
               "Do not use RobinsonTrautman analytic solution with a "
```
Cheers!

wthrowe commented:
Why is that different here?

moxcodes committed:
Fix occasional failures in RobinsonTrautman test

- tweak parameters to make run faster for fewer timeouts
- include `ExpectedOutput` so directories are cleaned up before
  re-running the test

moxcodes commented:
hm, you're right, the edge case was more peculiar. I don't properly understand the failure, because the case that caused a segfault in the `dynamic_cast`, when printed out, both of the versions were null pointers. However, it was my understanding that `dynamic_cast` was supposed to work fine with a null pointer (just returning a null pointer as well). And why `std::addressof` should work differently from `&` here is also mysterious now that I look into it further... 
The better fix was updating the test to actually construct the object with a generator so the dynamic cast is actually applied to a pointer to a polymorphic type:
```patch
diff --git a/src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp b/src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp
index db638051a0..761b6f4ba3 100644
--- a/src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp
+++ b/src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp
@@ -61,8 +61,8 @@ struct InitializeWorldtubeBoundaryBase {
     if constexpr (std::is_same_v<Tags::AnalyticBoundaryDataManager,
                                  tmpl::front<ManagerTags>>) {
       if (dynamic_cast<const Solutions::RobinsonTrautman*>(
-              std::addressof((db::get<Tags::AnalyticBoundaryDataManager>(box)
-                                  .get_generator()))) != nullptr) {
+              &(db::get<Tags::AnalyticBoundaryDataManager>(box)
+                    .get_generator())) != nullptr) {
         if(db::get<::Tags::TimeStepper<>>(box).number_of_substeps() != 1) {
           ERROR(
               "Do not use RobinsonTrautman analytic solution with a "
diff --git a/tests/Unit/Evolution/Systems/Cce/Actions/Test_InitializeWorldtubeBoundary.cpp b/tests/Unit/Evolution/Systems/Cce/Actions/Test_InitializeWorldtubeBoundary.cpp
index 1f593c7eab..634af40ac8 100644
--- a/tests/Unit/Evolution/Systems/Cce/Actions/Test_InitializeWorldtubeBoundary.cpp
+++ b/tests/Unit/Evolution/Systems/Cce/Actions/Test_InitializeWorldtubeBoundary.cpp
@@ -14,6 +14,7 @@
 #include "Domain/Structure/ElementId.hpp"
 #include "Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp"
 #include "Evolution/Systems/Cce/AnalyticSolutions/RobinsonTrautman.hpp"
+#include "Evolution/Systems/Cce/AnalyticSolutions/RotatingSchwarzschild.hpp"
 #include "Evolution/Systems/Cce/BoundaryData.hpp"
 #include "Evolution/Systems/Cce/Components/WorldtubeBoundary.hpp"
 #include "Evolution/Systems/Cce/InterfaceManagers/GhLockstep.hpp"
@@ -209,8 +210,10 @@ void test_analytic_initialization() noexcept {
       {l_max, 100.0, 0.0, std::make_unique<::TimeSteppers::RungeKutta3>()}};
 
   runner.set_phase(AnalyticMetavariables::Phase::Initialization);
-  ActionTesting::emplace_component<component>(&runner, 0,
-                                              AnalyticBoundaryDataManager{});
+  ActionTesting::emplace_component<component>(
+      &runner, 0,
+      AnalyticBoundaryDataManager{
+          12_st, 20.0, std::make_unique<Solutions::RotatingSchwarzschild>()});
   // this should run the initialization
   for (size_t i = 0; i < 3; ++i) {
     ActionTesting::next_action<component>(make_not_null(&runner), 0);
```
I also added a commit "Include ExpectedOutput in CCE Robinson-Trautman test" to fix a test failure that appeared on the most recent CI run.

Cheers!

moxcodes commented:
a test failed again, this time with a timeout on the RT test case, so I tweaked the parameters to make it run a bit more quickly. I guess now that it does self-start, it's a little slower because it needs to take more steps.
Cheers!

wthrowe reviewed (dismissed):
I don't see anything wrong with either version of the code, so I'm going to call it a compiler bug and not worry about it.

nilsdeppe commented on src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp:
```diff
@@ -57,6 +58,19 @@ struct InitializeWorldtubeBoundaryBase {
+          ERROR(
+              "Do not use RobinsonTrautman analytic solution with a "
+              "substep-based timestepper. This is to prevent severe slowdowns "
+              "in the current RobinsonTrautman implementation");
```
Is what these slowdowns are documented somewhere? If not, would it be useful to list them in the error message here so that if someone _really_ wants to use a substep integrator they can?

nilsdeppe reviewed (dismissed):
LGTM, just one question. If you decide to change anything, please squash immediately :)

moxcodes commented:
Added an additional note to the `RobinsonTrautman` solution better explaining the trouble with substep methods, and referred to that documentation in the error message.
Cheers!
